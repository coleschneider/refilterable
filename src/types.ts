import { History } from 'history';
import { MutableRefObject } from 'react';

export const CREATE_FILTER_MARKER = Symbol("createFilter");

export interface FilterConfig<T = string> {
	parse?(input: string): T; 
  format?(value: T): string; 
  validate?(input: string, parse: (input: string) => T): boolean;
  defaultValue?: T | undefined;
}

export interface FilterObject<T = string> extends FilterConfig<T> {
  paramName: string;
  parse(input: string): T; 
  format(value: T): string; 
  validate(input: string, parse: (input: string) => T): boolean;
}

/**
 * User defined type guard to check if the passed object is a valid instance of FiterObject
 * 
 * @param filter 
 */
export function isFilterObject<T>(filter: any): filter is FilterObject<T> {
  return filter[CREATE_FILTER_MARKER];
}

export type FiltersContextValue = {
  locationObserver: MutableRefObject<{ 
    watch(paramName: string, callback: Function): Function;
    getParamInfo(paramName: string): { hasParam: boolean, paramValue: any };
    getCurrentParams(): URLSearchParams;
  }>;
  history?: History;
  filterRegistry: Map<string, FilterObject<any>>,
}

export type ForwardHistoryAction = 'PUSH' | 'REPLACE';

export type SetFilterOptions = {
  dry?: boolean;
  action?: ForwardHistoryAction;
};

export type FilterSetter<T = string> = (nextValue: T, options?: SetFilterOptions) => string;
