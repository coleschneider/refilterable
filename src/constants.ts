import { SetFilterOptions, ForwardHistoryAction } from './types';

export const defaultSetFilterOptions: SetFilterOptions = Object.freeze({
  dry: false,
  action: 'PUSH' as ForwardHistoryAction,
});