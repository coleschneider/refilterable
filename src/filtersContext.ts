import { createContext } from "react";
import { FiltersContextValue } from './types';
import createLocationObserver from "./createLocationObserver";

export default createContext<FiltersContextValue | null>(null);