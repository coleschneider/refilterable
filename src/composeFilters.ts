import { FILTER_COMPOSITION_MARKER } from './utils/constants';
import { FilterObject, FilterComposition } from './utils/types';

// @todo: support filter compositions in filters
export default function composeFilters(filters: FilterObject[], validate?: (input: any) => boolean): FilterComposition {
  return ({
    filters,
    validate: typeof validate === 'function' ? validate : () => true,
    [FILTER_COMPOSITION_MARKER]: true,
  } as FilterComposition) ;
}