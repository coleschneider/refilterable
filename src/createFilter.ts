import invariant from 'invariant';
import { FilterConfig, FilterObject } from './utils/types';
import { CREATE_FILTER_MARKER } from './utils/constants';


export default function createFilter<T>(paramName: string, config?: FilterConfig<T>): FilterObject<T> {
  invariant(
    String(paramName || '').length > 0,
    `re-filter: param name cannot be empty`,
  );

  const defaultConfig = {
    parse(input: string): any { return input; }, 
    format(value: T): string { return String(value); },
    validate(): boolean { return true; },
    defaultValue: undefined,
    [CREATE_FILTER_MARKER]: true,
  };

  return {
    ...defaultConfig,
    ...config,
    paramName,
  };
}; 