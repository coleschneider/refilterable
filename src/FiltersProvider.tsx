import React, { useEffect, useRef, FunctionComponent } from 'react';
import invariant from 'invariant';
import { History, Location } from 'history';
import filtersContext from './filtersContext';
import { FiltersContextValue, FilterObject } from './types';
import createLocationObserver from './createLocationObserver';
 
type Props = {
  history: History;
  children: JSX.Element[];
}

const FiltersProvider: FunctionComponent<Props> = ({ history, children }: Props) => {
  invariant(
    history,
    `re-filter: FiltersProvider was not passed a history instance. 
     Make sure you pass it via the history prop`,
  );

  const filterRegistry = useRef<Map<string, FilterObject<any>>>(new Map());
  const locationObserver = useRef(createLocationObserver(history.location.search));

  useEffect(() => {
    // listen to location changes
    const unlisten = history.listen((location: Location) => {
      locationObserver.current.notify(location.search);
    });

    // unsubscribe when the component unmounts
    return unlisten;
  }, []);

  const contextValue: FiltersContextValue = { 
    locationObserver,
    history, 
    filterRegistry: filterRegistry.current 
  };

  return (
    <filtersContext.Provider value={contextValue}>
      {children}
    </filtersContext.Provider>
  );
}


export default FiltersProvider;