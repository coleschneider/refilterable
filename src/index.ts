import FiltersProvider from "./FiltersProvider";
import createFilter from "./createFilter";
import useFilter from "./useFilter";
import useReset from "./useReset";

export { useReset, useFilter, createFilter, FiltersProvider };
