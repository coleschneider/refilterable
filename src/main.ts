import * as React from 'react';
import FiltersProvider from "./FiltersProvider";
import createFilter from "./createFilter";
import useFilter from "./useFilter";
import useReset from "./useReset";

export default {
	FiltersProvider,
	createFilter,
	useFilter,
	useReset
};