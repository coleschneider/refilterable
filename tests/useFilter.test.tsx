import React from 'react';
import { History } from 'history';
import { render, act as reactTestingLibraryAct } from '@testing-library/react'
import { renderHook, act } from 'react-hooks-testing-library'
import { useFilter, FiltersProvider, createFilter, composeFilters } from '../src';
import { createMemoryHistory } from 'history';

describe('useFilter', () => {
  let wrapper, history: History;

  beforeEach(() => {
    history = createMemoryHistory();
    
    wrapper = ({ children }) => (
      <FiltersProvider 
        history={history}
      >
        {children}
      </FiltersProvider>
    );
  });

  beforeEach(() => {
    // suppress unncecessary warnings
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should throw an error when there is no provider in the tree', () => {
    expect(() => {
      renderHook(() => useFilter('foo')).result.current;
    }).toThrowErrorMatchingSnapshot();
  });

  it('should work if only the URL param name is passed', () => {
    const { result } = renderHook(() => useFilter<string>('foo'), { wrapper });;
    const [ foo, setFoo ] = result.current;

    expect(foo).toBeUndefined();
    expect(setFoo).toBeInstanceOf(Function);
  });

  it('should throw an error if there is a filter configuration collision', () => {
    const fooOne = createFilter('foo', { validate() { return true } });
    const fooTwo = createFilter('foo', { validate() { return false } });
    expect(() => {
      renderHook(() => [
        useFilter(fooOne),
        useFilter(fooTwo)
      ], { wrapper });
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      renderHook(() => [
        useFilter('foo'),
        useFilter(fooTwo)
      ], { wrapper });
    }).toThrowErrorMatchingSnapshot();
  });

  it('should not allow to initialize filters with plain objects', () => {
    const filter = JSON.parse(JSON.stringify(createFilter('foo')));

    expect(() => {
      renderHook(() => useFilter(filter), { wrapper }).result.current;
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      renderHook(() => useFilter({ paramName: undefined }), { wrapper }).result.current;
    }).toThrowErrorMatchingSnapshot();
  });

  describe('read and update filter values', () => {
    it('should allow to change the filter value', () => {
      const { result } = renderHook(() => useFilter('foo'), { wrapper });

      act(() => {
        const [_, setFoo] = result.current;
        setFoo('bar');
      });

      const [foo, _] = result.current;

      expect(foo).toBe('bar');
    });

    describe('parse()', () => {
      let parse, filter;

      beforeEach(() => {
        parse = jest.fn().mockImplementation(input => parseInt(input));
        filter = createFilter('foo', { parse });
      });

      it('should call the parse method in order to calculate the current value', () => {
        history.push({ search: '?foo=bar' });
        renderHook(() => useFilter(filter), { wrapper }).result.current;
        expect(parse).toHaveBeenCalledWith('bar');
      });
  
      it('should call the parse method with the default value when the filter is not specified', () => {
        history.push({ search: '?' });
        const filter = createFilter('foo', { parse, defaultValue: 'bar' });
        renderHook(() => useFilter(filter), { wrapper }).result.current;
        expect(parse).toHaveBeenCalledWith(filter.defaultValue);
      });

      it('should not call the parse method when the filter as well as its default value are not specified', () => {
        history.push({ search: '?' });
        renderHook(() => useFilter(filter), { wrapper }).result.current;
        expect(parse).not.toHaveBeenCalled();
      });

    });

    describe('validate()', () => {
      let parse, validate, filter;

      beforeEach(() => {
        parse = jest.fn().mockImplementation(input => parseInt(input));
        validate = jest.fn().mockImplementation((input, parse) => parse(input) >= 0);
        filter = createFilter('page', { parse, validate });
      });

      it('should call validate and return undefined if the value in the URL is invalid', () => {
        history.push({ search: '?page=-1' });
        const [page, _] = renderHook(() => useFilter(filter), { wrapper }).result.current;

        expect(validate).toHaveBeenCalled();
        expect(page).toBeUndefined();
      });

      it('should let the value be calculated if it is valid', () => {
        history.push({ search: '?page=1' });
        const [page, _] = renderHook(() => useFilter(filter), { wrapper }).result.current;

        expect(validate).toHaveBeenCalled();
        expect(page).toBe(1);
      });

    });

    describe('format()', () => {
      it('should throw an exception if a formatter produces a non-string value', () => {
        const format = jest.fn().mockReturnValue(100);
        const filter = createFilter('foo', { format });
        const [_, setFoo] = renderHook(() => useFilter(filter), { wrapper }).result.current;

        expect(() => act(() => {
          setFoo('bar');
        })).toThrowErrorMatchingSnapshot();
      });
    });

    describe('changing the filter value', () => {
      it('should reflect the new value in the URL', () => {
        const { result } = renderHook(() => useFilter('foo'), { wrapper });
        let nextSearch;

        act(() => {
          const [_, setFoo] = result.current;

          nextSearch = setFoo('bar');
        });

        const [foo, _] = result.current;
        expect(foo).toBe('bar');
        expect(history.location.search).toBe('?foo=bar');
        expect(nextSearch).toBe('foo=bar');
      });

      it('should not change the url if the "dry" option is set to true', () => {
        history.push({ search: 'foo=bar' });
        const { result } = renderHook(() => useFilter('foo'), { wrapper });
        let nextSearch;

        act(() => {
          const [_, setFoo] = result.current;
          nextSearch = setFoo('baz', { dry: true });
        });

        const [foo, _] = result.current;
        expect(foo).toBe('bar');
        expect(history.location.search).toBe('?foo=bar');
        expect(nextSearch).toBe('foo=baz');
      });

      it('should call history.push() if the action is not explicitly specified', () => {
        const push = jest.spyOn(history, 'push');
        const { result } = renderHook(() => useFilter('foo'), { wrapper });

        act(() => {
          const [_, setFoo] = result.current;
          setFoo('bar');
        });

        expect(push).toHaveBeenCalledWith({ search: 'foo=bar' })
      });

      it('should call history.replace() if the corresponding option is specified', () => {
        const replace = jest.spyOn(history, 'replace');
        const { result } = renderHook(() => useFilter('foo'), { wrapper });

        act(() => {
          const [_, setFoo] = result.current;
          setFoo('bar', { action: 'REPLACE' });
        });

        expect(replace).toHaveBeenCalledWith({ search: 'foo=bar' })
      });

    });

    describe('receiving chages from the outside', () => {
      it('should receive location changes (push)', () => {
        const { result } = renderHook(() => useFilter('foo'), { wrapper });

        act(() => {
          history.push({ search: 'foo=bar' });
        });

        const [foo, _] = result.current;

        expect(foo).toBe('bar');
      });

      it('should receive location changes (push)', () => {
        const { result } = renderHook(() => useFilter('foo'), { wrapper });

        act(() => {
          history.replace({ search: 'foo=bar' });
        });

        const [foo, _] = result.current;

        expect(foo).toBe('bar');
      });

      it('should call parse() and validate() when location changes', () => {
        const parse = jest.fn().mockReturnValue(10);
        const validate = jest.fn().mockReturnValue(true);
        const filter = createFilter('foo', { parse, validate })
        renderHook(() => useFilter(filter), { wrapper });

        act(() => {
          history.push({ search: 'foo=bar' });
        });

        expect(parse).toHaveBeenCalled();
        expect(validate).toHaveBeenCalled();
      });

      it('should not call parse() and validate() when location changes but it does not affect the current filter', () => {
        const parse = jest.fn().mockReturnValue(10);
        const validate = jest.fn().mockReturnValue(true);
        const filter = createFilter('foo', { parse, validate })
        renderHook(() => useFilter(filter), { wrapper });

        act(() => {
          history.push({ search: 'page=1000' });
        });

        expect(parse).not.toHaveBeenCalled();
        expect(validate).not.toHaveBeenCalled();
      });

    });

  });

  describe('interaction with provider', () => {
    it('should not allow multiple configurations to be attached to the same URL parameter', () => {
      expect(() => {
        renderHook(() => [
          useFilter<string>('foo'),
          useFilter<string>(createFilter('foo', { parse: (): string => '' })),
        ], { wrapper });
      }).toThrowErrorMatchingSnapshot();
    });

    it('should not throw an error when there are multiple configurations for the same URL parameter but under different providers', () => {
      expect(() => {
        renderHook(() => useFilter<string>('foo'), { wrapper });
        renderHook(() => (
          useFilter<string>(createFilter('foo', { parse: (): string => '' }))
        ), { wrapper });
      }).not.toThrow();
    });
  });

  describe('composite filters', () => {
    const minFilter = createFilter('min', { 
      parse: parseInt,
      defaultValue: '0',
    });
    const maxFilter = createFilter('max', { 
      parse: parseInt,
      defaultValue: '100',
    });
    
    const rangeFilter = composeFilters([
      minFilter, 
      maxFilter
    ], ({ min, max }) => min <= max);
    
    it('should take composite filters', () => {
      const { result } = renderHook(() => useFilter(rangeFilter), { wrapper });

      act(() => {
        history.push({ search: 'min=100&max=200' });
      });

      const [range] = result.current;

      expect(range).toEqual({ min: 100, max: 200 });
    });

    it('should allow to set composite filter values', () => {
      const { result } = renderHook(() => useFilter(rangeFilter), { wrapper });

      act(() => {
        const [, setRange] = result.current;
        setRange({ min: 100, max: 200 });
      });

      const [range] = result.current;

      expect(range).toEqual({ min: 100, max: 200 });
      expect(history.location.search).toBe('?max=200&min=100');
    });

    it('should allow to set values of the filters incrementally', () => {
      const { result } = renderHook(() => useFilter(rangeFilter), { wrapper });

      let nextSearch;

      act(() => {
        history.push({ search: 'min=100&max=200' });
        const [, setRange] = result.current;
        nextSearch = setRange({ min: 150 }, { incrementally: true });
      });

      const [range] = result.current;

      expect(range).toEqual({ min: 150, max: 200 });
      expect(nextSearch).toBe('max=200&min=150');
      expect(history.location.search).toBe('?max=200&min=150');
    });

    it('should not set values of the filters incrementally by default', () => {
      const { result } = renderHook(() => useFilter(rangeFilter), { wrapper });

      let nextSearch;

      act(() => {
        history.push({ search: 'min=100&max=200' });
        const [, setRange] = result.current;
        nextSearch = setRange({ min: 50 });
      });

      const [range] = result.current;

      expect(range).toEqual({ min: 50, max: 100 });
      expect(nextSearch).toBe('min=50');
      expect(history.location.search).toBe('?min=50');
    });

    it('should delete params when undefined is explicitly passed even if the "incrementally" flag is set to true', () => {
      const { result } = renderHook(() => useFilter(rangeFilter), { wrapper });

      let nextSearch;

      act(() => {
        history.push({ search: 'min=100&max=200' });
        const [, setRange] = result.current;
        nextSearch = setRange({ min: 50, max: undefined }, { incrementally: true });
      });

      const [range] = result.current;

      expect(range).toEqual({ min: 50, max: 100 });
      expect(nextSearch).toBe('min=50');
      expect(history.location.search).toBe('?min=50');
    });

    it('should return undefined if the values together are invalid', () => {
      const { result } = renderHook(() => useFilter(rangeFilter), { wrapper });

      let nextSearch;

      act(() => {
        history.push({ search: 'min=100&max=50' });
      });

      const [range] = result.current;
      expect(range).toBeUndefined();
    });
  });

  describe('performance', () => {
    let ConsumerComponent;
    let Wrapper;
    
    beforeEach(() => {
      Wrapper = wrapper;
      ConsumerComponent = jest.fn().mockImplementation(() => {
        const [foo] = useFilter('foo');
        return null;
      });
    });

    test('useFilter should cause the consumer component to rerender when critical values change (control group)', () => {
      render(
        <Wrapper>
          <ConsumerComponent />
        </Wrapper>
      );

      reactTestingLibraryAct(() => {
        history.push({ search: 'foo=bar' });
      });

      expect(ConsumerComponent).toHaveBeenCalledTimes(2);
    });

    test('useFilter should not cause the consumer component to rerender when irrelevant values change', () => {
      render(
        <Wrapper>
          <ConsumerComponent />
        </Wrapper>
      );

      reactTestingLibraryAct(() => {
        history.push({ search: 'bar=foo' });
      });

      expect(ConsumerComponent).toHaveBeenCalledTimes(1);
    });
  });

  it('should specify a proper hook debug value', () => {
    const useDebugValue = jest.spyOn(React, 'useDebugValue');
    history.push({ search: 'foo=bar' });
    renderHook(() => useFilter<string>('foo'), { wrapper });
    
    const debugCallback = useDebugValue.mock.calls[0][1];
    expect(useDebugValue).toHaveBeenCalledWith('useFilter', expect.any(Function));
    // @ts-ignore
    expect(debugCallback()).toBe('foo');
  });
});