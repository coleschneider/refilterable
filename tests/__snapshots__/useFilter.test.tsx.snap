// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useFilter interaction with provider should not allow multiple configurations to be attached to the same URL parameter 1`] = `"refilterable: you attempted to register conflicting configurations for the \\"foo\\" filter"`;

exports[`useFilter read and update filter values format() should throw an exception if a formatter produces a non-string value 1`] = `
"refilterable: a custom formatter (foo) produced a non-string value. 
            Make sure your formatter always returns a string"
`;

exports[`useFilter should not allow to initialize filters with plain objects 1`] = `
"refilterable: you called useFilter and passed an invalid filter object. 
      Instead of constructing the configuration object on your own, 
      use createFilter() or composeFilters()."
`;

exports[`useFilter should not allow to initialize filters with plain objects 2`] = `
"refilterable: you called useFilter and passed an invalid filter object. 
      Instead of constructing the configuration object on your own, 
      use createFilter() or composeFilters()."
`;

exports[`useFilter should throw an error if there is a filter configuration collision 1`] = `"refilterable: you attempted to register conflicting configurations for the \\"foo\\" filter"`;

exports[`useFilter should throw an error if there is a filter configuration collision 2`] = `"refilterable: you attempted to register conflicting configurations for the \\"foo\\" filter"`;

exports[`useFilter should throw an error when there is no provider in the tree 1`] = `"refilterable: Components that utilize the \\"useFilter\\" hook need to be wrapped with FiltersProvider."`;
